src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+    FieldOutput output;
+
+    if (step > shared_.minimum_step)
+    {
+        output = this->one_good_step(step, state);
+    }
+    else
+    {
+        // Do an integration step for a small step (a.k.a quick advance)
+        StepperResult result = stepper_(step, state);
+
+        // Update position and momentum
+        output.state = result.end_state;
+
+        real_type dyerr = truncation_error(
+            step, shared_.epsilon_rel_max, state, result.err_state);
+        output.step_taken = step;
+
+        // Compute a proposed new step
+        CELER_ASSERT(output.step_taken != 0);
+        output.value
+            = this->new_step_size(step, dyerr / (step * shared_.epsilon_step));
+    }
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Advance within a relative truncation error and estimate a good step size
+ * for the next integration.
+ */
+CELER_FUNCTION auto
+FieldDriver::one_good_step(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a proposed next step
+    FieldOutput output;
+
+    // Perform integration for adaptive step control with the trunction error
+    bool          condition       = false;
+    unsigned int  remaining_steps = shared_.max_nsteps;
+    real_type     errmax2 = celeritas::numeric_limits<real_type>::max();
+    StepperResult result;
+
+    do
+    {
+        result  = stepper_(step, state);
+        errmax2 = truncation_error(
+            step, shared_.epsilon_rel_max, state, result.err_state);
+
+        if (errmax2 <= 1)
+        {
+            condition = true;
+        }
+        else
+        {
+            // Step failed; compute the size of re-trial step.
+            real_type htemp = shared_.safety * step
+                              * std::pow(errmax2, half() * shared_.pshrink);
+
+            // Truncation error too large, reduce stepsize with a low bound
+            step = std::fmax(htemp, shared_.max_stepping_decrease * step);
+        }
+    } while (!condition && --remaining_steps > 0);
+
+    // TODO: loop check and handle rare cases if happen
+    CELER_ASSERT(condition);
+
+    // Update state, step taken by this trial and the next predicted step
+    output.state      = result.end_state;
+    output.step_taken = step;
+    output.value      = (errmax2 > ipow<2>(shared_.errcon))
+                       ? shared_.safety * step
+                             * std::pow(errmax2, half() * shared_.pgrow)
+                       : shared_.max_stepping_increase * step;
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ *  Estimate the new predicted step size based on the error estimate
+ */
+CELER_FUNCTION
+real_type FieldDriver::new_step_size(real_type step, real_type error) const
+{
+    CELER_ASSERT(error > 0);
+    real_type scale_factor = (error > 1) ? std::pow(error, shared_.pshrink)
```

Comment:
This (`> 1`) seems to indicate that this `error` is normalized (to something).  Are you the error passed normalized or are some in absolute value? (If the later, consider changing the name of one or the other to indicate absolute vs normalized).

Reply:
Improved readability (error --> rel_error).  Probably, the relative error ratio is the better name, but rel_error is taken to keep it to be short.